File: backend/voting-app/src/main/java/com/tu/votingapp/entities/referendum/ReferendumVoteEntity.java
package com.tu.votingapp.entities.referendum;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(
        name = "referendum_votes",
        uniqueConstraints = @UniqueConstraint(
                columnNames = {"user_id", "referendum_id"}
        )
)
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ReferendumVoteEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Voter's ID from the UserPrincipal.
     */
    @Column(name = "user_id", nullable = false)
    private Long userId;

    /**
     * Referendum in which this vote was cast.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "referendum_id", nullable = false)
    private ReferendumEntity referendum;

    /**
     * Option chosen by the voter.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "option_id", nullable = false)
    private ReferendumOptionEntity option;

    /**
     * Timestamp when the vote was cast.
     */
    @Column(name = "vote_timestamp", nullable = false)
    private LocalDateTime voteTimestamp;
}

File: frontend/android/app/src/main/java/com/example/android/entity/referendum/ReferendumVoteEntity.kt
package com.example.android.entity.referendum

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity(tableName = "referendum_votes")
data class ReferendumVoteEntity(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    val id: Long? = null,

    @ColumnInfo(name = "user_id")
    val userId: Long,

    @ColumnInfo(name = "referendum_id")
    val referendumId: Long,

    @ColumnInfo(name = "option_id")
    val optionId: Long,

    @ColumnInfo(name = "vote_timestamp")
    val voteTimestamp: String // ISO 8601 format, e.g. "2025-04-20T14:22:00"
)